src/main/java/com/zepben/cimbend/get/DiagramConsumerClient.kt

```diff
@@ -32,54 +34,63 @@ class DiagramConsumerClient(
      *
-     * @return A [Map] containing the retrieved objects keyed by mRID. If an item is not found it will be excluded from the map.
+     * WARNING: This operation is not atomic upon [service], and thus if processing fails partway through [mRIDs], any previously successful mRID will have been
+     * added to the service, and thus you may have an incomplete [DiagramService]. Also note that adding to the [service] may not occur for an object if another
+     * object with the same mRID is already present in [service]. [MultiObjectResult.failed] can be used to check for mRIDs that were retrieved but not
+     * added to [service].
+     *
+     * @return A [GrpcResult] with a result of one of the following:
+     * - A [MultiObjectResult] containing a map of the retrieved objects keyed by mRID. If an item is not found it will be excluded from the map.
+     *   If an item couldn't be added to [service] its mRID will be present in [MultiObjectResult.failed] (see [com.zepben.cimbend.common.BaseService.add]).
+     * - A [Throwable] if an error occurred while retrieving or processing the objects, in which case, [GrpcResult.wasSuccessful] will return false.
+     * Note the warning above in this case.
      */
-    override fun getIdentifiedObjects(
-        service: DiagramService,
-        mRIDs: Iterable<String>
-    ): GrpcResult<Map<String, IdentifiedObject>> {
+    override fun getIdentifiedObjects(service: DiagramService, mRIDs: Iterable<String>): GrpcResult<MultiObjectResult> {
         return tryRpc {
-            processIdentifiedObjects(service, GetIdentifiedObjectsRequest.newBuilder().addAllMrids(mRIDs).build())
-                .filterNotNull()
-                .associateBy({ it.mRID }, { it })
+            processIdentifiedObjects(service, mRIDs).let { extracted ->
+                val results = mutableMapOf<String, IdentifiedObject>()
+                val failed = mutableSetOf<String>()
+                extracted.forEach {
+                    if (it.identifiedObject == null) failed.add(it.mRID) else results[it.identifiedObject.mRID] = it.identifiedObject
+                }
+                MultiObjectResult(results, failed)
+            }
         }
     }
 
 
-    private fun processIdentifiedObjects(
-        service: DiagramService,
-        request: GetIdentifiedObjectsRequest
-    ): Sequence<IdentifiedObject?> {
-        return stub.getIdentifiedObjects(request)
-            .asSequence()
-            .flatMap { it.identifiedObjectsList.asSequence() }
-            .map { extractIdentifiedObject(service, it) }
-    }
+    private fun processIdentifiedObjects(service: DiagramService, mRIDs: Iterable<String>): Sequence<ExtractResult> =
+        mRIDs.filter { service.get<IdentifiedObject>(it) == null }.let {  // Only process mRIDs not in the service
+            stub.getIdentifiedObjects(GetIdentifiedObjectsRequest.newBuilder().addAllMrids(mRIDs).build())
```

Comment:
Likewise this should not be passing in mRIDs to addAllMrids. It should be addAllMrids(it).